executions/graphql-kotlin-dataloader-instrumentation/src/main/kotlin/com/expediagroup/graphql/dataloader/instrumentation/level/DataLoaderLevelInstrumentation.kt

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2022 Expedia, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.expediagroup.graphql.dataloader.instrumentation.level
+
+import com.expediagroup.graphql.dataloader.KotlinDataLoaderRegistry
+import com.expediagroup.graphql.dataloader.instrumentation.level.execution.AbstractExecutionLevelInstrumentation
+import com.expediagroup.graphql.dataloader.instrumentation.level.execution.OnLevelDispatched
+import com.expediagroup.graphql.dataloader.instrumentation.level.execution.ExecutionLevelInstrumentationParameters
+import com.expediagroup.graphql.dataloader.instrumentation.level.state.Level
+import graphql.ExecutionInput
+import graphql.GraphQLContext
+import graphql.execution.instrumentation.Instrumentation
+import org.dataloader.DataLoader
+
+/**
+ * Custom GraphQL [Instrumentation] that will dispatch all [DataLoader]s inside a [KotlinDataLoaderRegistry]
+ * when certain [Level] is dispatched for all [ExecutionInput] sharing a [GraphQLContext]
+ */
+class DataLoaderLevelInstrumentation : AbstractExecutionLevelInstrumentation() {
+    override fun calculateLevelDispatchedState(
```

Comment:
you still have a `calculateLevelDispatchedState` function that returns `OnLevelDispatched` which is somewhat confusing....

Reply:
changed the approach to use simple typed callbacks instead of interfaces, that way we could do

```kotlin
class DataLoaderSyncExecutionExhaustedInstrumentation : AbstractSyncExecutionExhaustedInstrumentation() {
    override fun getOnSyncExecutionExhaustedCallback(
        parameters: SyncExecutionExhaustedInstrumentationParameters
    ): OnSyncExecutionExhaustedCallback = {
        parameters
            .executionContext
            .graphQLContext.get<KotlinDataLoaderRegistry>(KotlinDataLoaderRegistry::class)
            ?.dispatchAll()
    }
}
```

where `OnSyncExecutionExhaustedCallback`

is 
```kotlin
internal typealias OnSyncExecutionExhaustedCallback = (List<ExecutionInput>) -> Unit
```